Retry obs_reset_video with default params if user ones fail

eugen15 opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

### Description
- Moved `obs_reset_video` call to a separate wrapper method
- Extended the `OBS_service::resetVideoContext` method to check for errors and retry initialization with default video configuration if the user one fails
- Added the `OBS_service::keepFallbackVideoConfig` method to copy the default video configuration to the user one, then save it to `basic.ini`; the method is used by `OBS_service::resetVideoContext` if retrying is successful

### Motivation and Context
Theoretically, this change may help when a user has a corrupted config or wrong settings.

### How Has This Been Tested?
1. Change one of the `basic.ini` settings to -1 or 65535
```
[Video]
BaseCX=1920
BaseCY=1080
OutputCX=1920
OutputCY=1080
```
2. Change one of non-**Video** settings to something non-default but standard (i.e. a correct value)
3. Start Streamlabs Desktop (without the changes it will not start correctly)
4. Check logs in `slobs-client\node-obs\logs` for the following messages:
- `Prepared obs_video_info:` for the first occurrence it has the incorrect value; for the other ones it has a default value
- `About to reset the video context with the user configuration`
- `The video context reset with the user configuration failed: -3`
- `About to reset the video context with the default configuration`
- `Saving the fallback/default video configuration to basic.ini`
5. `basic.ini` has the default **Video** config
6. Check that non-Video parameters has not changed.

Tested on Windows and macOS.

### Types of changes
A theoretical bug fix.

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] Some comments added to the code.

eugen15 committed:
Retry obs_reset_video with default params if user ones fail

EddyGharbi commented on obs-studio-server/source/nodeobs_service.cpp:
```diff
@@ -442,9 +442,44 @@ int OBS_service::resetVideoContext(bool reload)
+{
 	try {
-		return obs_reset_video(&ovi);
+		return obs_reset_video(const_cast<obs_video_info*>(&ovi));
```
Why do we need to cast here?

eugen15 replied:
Fixed. Explained in a code comment.

EddyGharbi commented on obs-studio-server/source/nodeobs_service.h:
```diff
@@ -293,12 +293,17 @@ class OBS_service
 	// Reset contexts
 	static bool resetAudioContext(bool reload = false);
 	static int  resetVideoContext(bool reload = false);
+	static int	doResetVideoContext(const obs_video_info& ovi);
```
Extra spaces

EddyGharbi commented on obs-studio-server/source/nodeobs_service.cpp:
```diff
@@ -442,9 +442,44 @@ int OBS_service::resetVideoContext(bool reload)
+	//   libobs-opengl.
+	// OBS_VIDEO_FAIL: Generic failure.
+	if (errorcode == OBS_VIDEO_FAIL || errorcode == OBS_VIDEO_INVALID_PARAM) {
+
```
Extra line

EddyGharbi commented on obs-studio-server/source/nodeobs_service.cpp:
```diff
@@ -442,9 +442,44 @@ int OBS_service::resetVideoContext(bool reload)
+	// OBS_VIDEO_MODULE_NOT_FOUND: Could not load a dynamic library (ovi.graphics_module):
+	//   libobs-d3d11.dll,
+	//   libobs-opengl.
+	// OBS_VIDEO_FAIL: Generic failure.
```
Thanks for this, good call to add it as comments, very clear

eugen15 committed:
Minor fixes; limited the retry to the first init call only

Fixed formatting/removed extra spaces; added the retryWithDefaultConf parameter (false by default) to resetVideoContext; removed the const_cast from obs_reset_video call; limited "retry" to the only path: OBS_API::OBS_API_initAPI -> OBS_API::openAllModules -> OBS_service::resetVideoContext
